controllers/ingressnodefirewallconfig_controller.go

```diff
@@ -0,0 +1,137 @@
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.12.1/pkg/reconcile
+func (r *IngressNodeFirewallConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	logger := r.Log.WithValues("ingress node firewall config", req.NamespacedName)
+
+	instance := &ingressnodefwv1alpha1.IngressNodeFirewallConfig{}
+	err := r.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	if req.Name != defaultIngressNodeFirwallCrName {
+		logger.Error(err, "Invalid IngressNode firewall config resource name", "name", req.Name)
+		return ctrl.Result{}, nil // Return success to avoid requeue
+	}
+
+	if err := r.syncIngressNodeFwConfigResources(instance); err != nil {
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *IngressNodeFirewallConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&ingressnodefwv1alpha1.IngressNodeFirewallConfig{}).
+		Complete(r)
+}
+
+func (r *IngressNodeFirewallConfigReconciler) syncIngressNodeFwConfigResources(config *ingressnodefwv1alpha1.IngressNodeFirewallConfig) error {
+	logger := r.Log.WithName("syncIngressNodeFirewallConfigResources")
+	logger.Info("Start")
+	data := render.MakeRenderData()
+
+	data.Data["Image"] = os.Getenv("DAEMONSET_IMAGE")
+	objs, err := render.RenderDir(ManifestPath, &data)
+	if err != nil {
+		logger.Error(err, "Fail to render config daemon manifests")
+		return err
+	}
+
+	for _, obj := range objs {
+		if obj.GetKind() == "DaemonSet" &&
+			(len(config.Spec.NodeSelector) > 0 || len(config.Spec.Tolerations) > 0) {
+			scheme := kscheme.Scheme
+			ds := &appsv1.DaemonSet{}
+			err = scheme.Convert(obj, ds, nil)
+			if err != nil {
+				logger.Error(err, "Fail to convert IngressNodeFirewallConfig object to DaemonSet")
+				return err
+			}
+			if len(config.Spec.NodeSelector) > 0 {
+				ds.Spec.Template.Spec.NodeSelector = config.Spec.NodeSelector
+			}
+			if len(config.Spec.Tolerations) > 0 {
+				ds.Spec.Template.Spec.Tolerations = config.Spec.Tolerations
+			}
+			err = scheme.Convert(ds, obj, nil)
+			if err != nil {
+				logger.Error(err, "Fail to convert DaemonSet to IngressNodeFirewallConfig object")
+				return err
+			}
+		}
+
+		if err := apply.ApplyObject(context.TODO(), r.Client, obj); err != nil {
```

Comment:
The same here: ctx instead of context.TODO()?

Reply:
Move apply.ApplyObject into the previous ifcondition (line 130), otherwise you apply non-deamonsets or null objects from` ./bindata/manifests/daemon/daemonset.yaml `

------

When you define, then the file parses into 2 objects (one daemonset, one null object):
~~~
[akaris@linux ingress-node-firewall (daemon)]$ cat ./bindata/manifests/daemon/daemonset.yaml 

---
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: ingr
~~~

You run apply.ApplyObject regardless of if this matches:
~~~
		if obj.GetKind() == "DaemonSet" &&
			(len(config.Spec.NodeSelector) > 0 || len(config.Spec.Tolerations) > 0) {
~~~

Reply:
Even after correcting that, I get:
~~~
"could not apply (apps/v1, Kind=DaemonSet) ingress-node-firewall-daemon: could not update object (apps/v1, Kind=DaemonSet) /ingress-node-firewall-daemon: an empty namespace may not be set when a resource name is provided",
~~~